src/server/ua_services_attribute.c

```diff
@@ -992,12 +992,17 @@ UA_Boolean
 compatibleValueArrayDimensions(const UA_Variant *value, size_t targetArrayDimensionsSize,
                                const UA_UInt32 *targetArrayDimensions) {
     size_t valueArrayDimensionsSize = value->arrayDimensionsSize;
-    UA_UInt32 *valueArrayDimensions = value->arrayDimensions;
+    UA_UInt32 const *valueArrayDimensions = value->arrayDimensions;
     UA_UInt32 tempArrayDimensions;
     if(!valueArrayDimensions && !UA_Variant_isScalar(value)) {
-        valueArrayDimensionsSize = 1;
-        tempArrayDimensions = (UA_UInt32)value->arrayLength;
-        valueArrayDimensions = &tempArrayDimensions;
+        if(value->arrayLength > 0) {
+            valueArrayDimensionsSize = 1;
+            tempArrayDimensions = (UA_UInt32)value->arrayLength;
+            valueArrayDimensions = &tempArrayDimensions;
+        } else {
```

Comment:
This always accepts arrays of length zero with no array dimensions.
Here, we should always check that `dimSum==0` as well.
Otherwise we can slip in an array of length zero even if the value rank requires something else.